### PR TITLE
Add support for customer-defined partition keys

### PR DESCRIPTION
**Pull request by trevorbonas**
- Users can now define partition keys for their newly-created tables. Partition key configuration is controlled using three environment variables, `custom_partition_key_type`, `custom_partition_key_dimension`, and `enforce_custom_partition_key`.
- Environment variables added for CDPK support.
- SAM template parameters added for CDPK support.
- Documentation added for CDPK support.
- Integration tests added for CDPK support.
- Integration test asserts moved to end of tests, in order to ensure resources are cleaned up.

- [x] Unit tests passed.
- [x] Integration tests passed.

### CODE REVIEW

**Comment by trevorbonas** [comment]
This PR looks like it has a lot of changes but that is because I pulled in the changes from `dev-sam-template` in order to add custom-defined partition key support to the SAM template. Once that branch has been merged the number of changes in this PR will reduce.

File: integrations/influxdb_connector/influxdb-timestream-connector/README.md [hunk 1]
```diff
@@ -55,10 +55,18 @@ The following parameters are available when deploying the connector as part of a
 
 | Parameter     | Description | Default Value |
 |---------------|-------------|---------------|
+| `CustomPartitionKeyDimension` |  The dimension to use as the partition key. This parameter is required if the CustomPartitionKeyType parameter is set to `dimension`. | |
+| `CustomPartitionKeyType` | The type of custom partition key to use. Valid options are `dimension` or `measure`. The `dimension` option requires the CustomPartitionKeyDimension parameter to also be set. If this parameter is not provided, newly-created tables will use default partitioning and none of the parameters relating to custom partition keys will be used. | |
```
Comment (MaxKsyunz) [inline integrations/influxdb_connector/influxdb-timestream-connector/README.md:59]: > If this parameter is not provided, ...
Is this behavior useful and expected? The previous sentence implies that missing `dimension` should result in a validation error.

Comment (trevorbonas) [reply integrations/influxdb_connector/influxdb-timestream-connector/README.md:59]: Whether `CustomPartitionKeyType` has a value determines whether the other partition key environment variables are used or ignored.

`dimension` is an "option" in the sense that it is one possible value for `CustomPartitionKeyType`. Depending on whether `CustomPartitionKeyType` is `dimension` or `measure`, other environment variables need to be set.

Comment (MaxKsyunz) [reply integrations/influxdb_connector/influxdb-timestream-connector/README.md:59]: What happens if the other variables are not set?

Comment (trevorbonas) [reply integrations/influxdb_connector/influxdb-timestream-connector/README.md:59]: If `CustomPartitionKeyType` is `measure`, no other environment variables need to be set. If `CustomPartitionKeyType` is `dimension` then we do a check of these other environment variables and return an `Err` if any are missing.

If we didn't do a check, we would eventually get an error when attempting to create a table.

I noticed that the error messages in one of our checks is inaccurate, I have fixed this.

**Review by MaxKsyunz — APPROVED** [review]

**Pull request merged**